### PR TITLE
0ttl

### PR DESCRIPTION
**Pull request by Philip-NLnetLabs**
Fix for issue #860

### CODE REVIEW

File: services/cache/dns.c [hunk 1]
```diff
@@ -183,12 +178,18 @@ dns_cache_store_msg(struct module_env* env, struct query_info* qinfo,
 		 * which could be useful for delegation information */
 		verbose(VERB_ALGO, "TTL 0: dropped msg from cache");
 		free(rep);
-		/* if the message is SERVFAIL in cache, remove that SERVFAIL,
+		/* if the message is in then cache, remove that msg,
```
Comment (wcawijngaards) [inline services/cache/dns.c]: spelling fix.
```suggestion
		/* if the message is in the cache, remove that msg,
```

File: testdata/serve_expired_0ttl_nodata.rpl [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+stub-zone:
+	name: "example.com"
+	stub-addr: 1.2.3.4
+CONFIG_END
+
+SCENARIO_BEGIN Test serve-expired with NXDOMAIN followed by 0 TTL
+; Scenario overview:
+; - query for 0ttl.example.com. IN A
+; - answer from upstream is NODATA; will be cached for NORR_TTL(5)
```
Comment (wcawijngaards) [inline testdata/serve_expired_0ttl_nodata.rpl]: Fix TTL value in comment.
```suggestion
; - answer from upstream is NODATA; will be cached for the SOA negative TTL.
```

File: testdata/serve_expired_0ttl_nxdomain.rpl [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+stub-zone:
+	name: "example.com"
+	stub-addr: 1.2.3.4
+CONFIG_END
+
+SCENARIO_BEGIN Test serve-expired with NXDOMAIN followed by 0 TTL
+; Scenario overview:
+; - query for 0ttl.example.com. IN A
+; - answer from upstream is NXDOMAIN; will be cached for NORR_TTL(5)
```
Comment (wcawijngaards) [inline testdata/serve_expired_0ttl_nxdomain.rpl]: Fix TTL value in comment.
```suggestion
; - answer from upstream is NXDOMAIN; will be cached for the SOA negative TTL.
```

**Review by wcawijngaards — APPROVED** [review]
I think it is a good idea to evict the message cache content when a 0 TTL new response is received. That synchronizes the content to the upstream content. The code looks good to have.

**Review by gthess — CHANGES_REQUESTED** [review]
Looks good!
But I believe we don't need the `msg_del_for_0ttl()` function anymore, just straight `msg_cache_remove()` in that place.
I would also like to see the same behavior for the ECS cache and cachedb.

**Comment by Philip-NLnetLabs** [comment]
Yes, I'll change it to call msg_cache_remove. Maybe you can to the other two issues in a separate ticket.

**Pull request merged**